e2e-tests/helpers/boundary-api-client.ts

```diff
@@ -0,0 +1,205 @@
+        patchEmptyResponsesMiddleware,
+      ],
+    });
+
+    return openapiConfiguration;
+  }
+
+  deleteMethod(
+    resourceType: string,
+  ):
+    | ((
+        args: { id: string },
+        initOverrides?: RequestInit | InitOverrideFunction,
+      ) => Promise<object>)
+    | undefined {
+    const { clients } = this;
+
+    const deleteMethods = {
+      roles: clients.roles.roleServiceDeleteRole.bind(clients.roles),
+      scopes: clients.scopes.scopeServiceDeleteScope.bind(clients.scopes),
+      targets: clients.targets.targetServiceDeleteTarget.bind(clients.targets),
+    };
+
+    return deleteMethods[resourceType];
+  }
+
+  skipCleanup(identifiable: { id: string }): void {
+    this.skipCleanupResources.push(identifiable);
+  }
+
+  shouldSkipCleanup(identifiable: { id: string }): boolean {
+    return this.skipCleanupResources.some(
+      (resource) => resource.id === identifiable.id,
+    );
+  }
+}
+
+// the openapi spec generated incorrectly specifies that POST responses will have a json body,
+// but they actually return a 204 "No Content" status with no body. The generated api client
+// does not handle this correctly, so we need to patch it here when it tries to access the `json`
+// method on the response
+const patchEmptyResponsesMiddleware = {
+  async post(context: BoundaryApiClient.ResponseContext) {
+    if (context.response.status === 204 && context.response.body === null) {
+      context.response.json = async () => undefined;
+    }
+
+    return context.response;
+  },
+};
+
+const betterErrorHandlingMiddleware = {
+  async post(context: BoundaryApiClient.ResponseContext) {
+    // skip logging of cleanup requests
+    if (context.init.headers?.[BoundaryE2EResourceCleanupHeader] === 'true') {
+      return;
+    }
+
+    if (context.response.status >= 400 && context.response.status < 600) {
+      let errorResponseMessage;
+      const clonedResponse = context.response.clone();
+      try {
+        errorResponseMessage = await clonedResponse.text();
+      } finally {
+        console.error(
+          `Received status code: ${clonedResponse.status}, with method: ${context.init.method}, on ${clonedResponse.url}${errorResponseMessage ? `:\n${errorResponseMessage}` : ''}`,
+        );
+      }
+    }
+    return context.response;
+  },
+};
+
+function extractDomainResourceFromUrl(url: string): string | undefined {
+  const match = new URL(url).pathname.match(/\/v1\/([a-z]+)(|\/.*)/);
+  if (!match) return undefined;
+
+  const [_, type] = match;
+  return type;
+}
+
+export const boundaryApiClientTest = base.extend<{
+  apiClient: BoundaryApi;
+  controllerAddr?: string;
+}>({
+  apiClient: async ({ request, controllerAddr }, use) => {
+    if (!controllerAddr) {
+      throw new Error('`controllerAddr` must be set in the base fixture');
+    }
+
+    const boundaryApi = new BoundaryApi(controllerAddr);
+    await use(boundaryApi);
+
+    for (const [resourceType, resources] of Object.entries(
```

Comment:
Very slick, I love how this is handled for resource cleanup by directly integrating with playwright